projects/components/src/table/data/table-cdk-data-source.ts

```diff
@@ -328,11 +328,19 @@ export class TableCdkDataSource implements DataSource<TableRow> {
 
   private paginateRows(rows: TableRow[], pageConfig: PageEvent): TableRow[] {
     /*
-     * The "rows" here are the results that are fetched. Since they are fetched with an offset in the request, we just
-     * index off the start of the result rows.
+     * With server-side pagination, the "rows" here are the results that are fetched. Since they are fetched with an
+     * offset in the request, we just index off the start of the result rows.
+     *
+     * However for when the backend APIs do not support pagination, the backend will return all results. In this case
+     * we can see if there are more results than we requested. If so, paginate within those results here.
+     *
+     * Historical Note & Warning:
+     * In the distant past for even server-side pagination we would prefetch additional rows, cache them, and
+     * paginate within those results. This caching will no longer work. In fact, it will break the below logic
```

Comment:
This seems fairly easily fixable though, which would be an additional win. If we pass the request into this, we have the offset and thus know what the data's start is too, right?

Reply:
I like this, but not sure it will work. The table is the one that generates the request - it applies the limits, sorting, filtering, etc based on the usage of the table component and paginator.

Reply:
I agree, that alone won't work here - but in broad strokes, let me run through how it used to work before refactor.

The table generates the requests, sends it to the data source, who we presume to respect the request. Whether it does that client or server side is not a concern of the table. So if we sent a request with offset 0 and limit 50, we expect the first row back to be the first result at the request offset.

So if we get 100 results back, we should be able to slice that into 0-50 for display, easy enough. Now, what's missing is that when we get a new pagination event, rather than blindly firing a new request we used to check our current rows and the request that generated them. If we're just requesting the next page and the prior result had enough rows to satisfy that, we can just reslice and be on our way. If not, we can fire a query like we are now.